src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -131,6 +132,53 @@ public void copyFrom(MemorySegment src) {
                 base(), min(), size);
     }
 
+    @Override
+    public long mismatch(MemorySegment other) {
+        AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)other;
+        final long thisSize = this.byteSize();
+        final long thatSize = that.byteSize();
+        final long minSize = Math.min(thisSize, thatSize);
+
+        this.checkRange(0, minSize, false);
+        that.checkRange(0, minSize, false);
+
+        if (this == other)
+            return -1;
+
+        long off = 0;
+        long remaining = minSize;
+        int i = 0;
+        while (remaining > 7) {
+            int size;
+            if (remaining > Integer.MAX_VALUE) {
+                size = Integer.MAX_VALUE;
+            } else {
+                size = (int) remaining;
+            }
+            i = ArraysSupport.vectorizedMismatch(
+                    this.base(),
+                    this.min() + off,
+                    that.base(),
+                    that.min() + off,
+                    size,
+                    ArraysSupport.LOG2_ARRAY_BYTE_INDEX_SCALE);
+            if (i >= 0) {
+                return off + i;
+            }
+            i = size - ~i;
+
+            off += i;
+            remaining -= i;
+        }
+
+        for (; off < minSize; off++) {
+            if (UNSAFE.getByte(this.base(), this.min() + off) != UNSAFE.getByte(that.base(), that.min() + off)) {
```

Comment:
This code could be simplified/rewritten to use `MemoryAddress' and VH, instead of unsafe access with object/offset addressing. E.g. you could maintain a `MemoryAddress` local variable in the loop instead of the `offset` and keep increasing that address on each iteration of `ArraySupport::vectorizedMismatch`. Then, when you get out of the loop, the address already points at the base of the region to compare, and a simple for loop with an indexed VH should do the job.